docs/RFCS/20221124_split_tls.md

```diff
@@ -0,0 +1,295 @@
+- Feature Name: Split service TLS certificates
+- Status: draft
+- Start Date: 2022-11-24
+- Authors: knz catj
+- RFC PR: [#92422](https://github.com/cockroachdb/cockroach/pull/92499)
+- Cockroach Issue: [#39326](https://github.com/cockroachdb/cockroach/issues/39326) [#92069](https://github.com/cockroachdb/cockroach/issues/92069)
+
+# Summary
+
+
+This RFC proposes to use a separate server certificate for the
+SQL/HTTP/RPC services for _clients_, i.e. non-node peers.
+
+This change has two main motivations:
+- it enables uses of a separate PKI for the interaction between client
+  apps/LBs and CockroachDB, from the one used internally between
+  nodes. This is a strong ask from Enterprise customers.
+- it also enables use of separate server hosts/IP addresses as seen
+  from (and validated by) clients, from the one seen from / validated
+  by peer nodes. This was found useful for certain types of user
+  orchestration.
+
+The proposal is twofold:
+- in the **simple and recommended deployment style**, we will tell users
+  to use two certificates: an _internal_ certificate for node-node
+  connections (current `node.crt`), and an _external_ certificate (new:
+  `server.crt`) to present to incoming non-node connections: HTTP API
+  clients, SQL clients and RPC admin CLI commands.
+- for **advanced deployments only** (not recommended as starting point in
+  docs), where a user has a legitimate need for customization of which
+  cert is used for which service, we will allow _per-service
+  certificates_: the current `ui.crt` for HTTP, and `server.crt` for
+  the SQL/RPC services.
+
+In any case, we will change CockroachDB to **stop using `node.crt` to
```

Comment:
Even if we automatically copy node.crt to server.crt on upgrade, it is still a breaking change for automation that creates nodes or pushes certificates. Since the automation would need to be enhanced to support generating server.crt. Instead of copying the certificate, I would be in favor of falling back on the node.crt if the server.crt is not present.

The fall back behavior is more complex, but we can document the behavior as certificate preferences. 

```
node to node: node.crt
sql server: server.crt > node.crt
http server: ui.crt > server.crt > node.crt
```

Here is an example bug the breaking change could introduce: A customer has automation that periodically pushes new certificates to the node. On upgrade, cockroach db copies node.crt to server.crt. A few weeks later the automation pushes a new node.crt. Since the automation does not know of server.crt, it is never updated and eventually expires, causing an outage.

Reply:
I'd argue the fallback behaviour is a large part of why our PKI code is frustrating to work with. I'd agree that automation will need to be updated to match this, but that change can be mitigated with operating system level fixes (symbolically linking `server.crt` and `ui.crt` to `node.crt`, a separate step in the automation that copies files after automation runs, etc). Ideally we want the `node.crt` certificates to be managed separately from `server.crt`, since that opens the possibility of two different pieces of automation, one cluster level and one externally facing.

Reply:
To clarify what I mean about the fallback being frustrating to work with: we simultaneously have to support `node.crt` and `ca.crt` being the fallback for every client/server certificate and certificate authority without requiring the use of any sort of field that we can then leverage for validation (since we have to support `node.crt` and `ca.crt` being issued off of an external PKI that may not be able to set custom fields or may be operated by people who don't want to set custom fields). Creating a clear distinction between `node.crt` and `server.crt` allows us to improve automation related to both, since we no longer have to treat them as the same certificate, while allowing existing users to use the same certificate if their PKI supports them doing so.